rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,620 @@
+        return tf.cast(keys % shard_num, dtype=tf.int32)
+      ```
+
+      shared_name: If non-empty, the SparseVariable will be shared under
+        the given name across multiple sessions.
+      name: A name for the operation (optional).
+      initial_value: The value to use if a key is missing in the hash table.
+      initial_mode: define the behavior when some keys were missing.
+        'random': lookup will return the random values
+          with normal distribution(mean=0.0, stddev=0.1)
+        'constant': lookup will return the valuse
+          filled by `initial_value`.
+      trainable: True, will be treated as a trainable Variable, and add to
+        to the list of variables collected in the graph under the key
+        `GraphKeys.TRAINABLE_VARIABLES`.
+      checkpoint: if True, the contents of the Variable are
+        saved to and restored from checkpoints.
+        If `shared_name` is empty for a checkpointed table,
+        it is shared using the table node name.
+
+    Returns:
+      A `dynamic_embedding.Variable` object.
+    """
+    self.key_dtype = key_dtype
+    self.value_dtype = value_dtype
+    self.default_value = initial_value
+    self.dim = dim
+    self.devices = devices if isinstance(devices, list) else [devices, ]
+    self.partition_fn = partitioner
+    self.name = name
+    self.shared_name = shared_name or "shared_name.{}".format(name)
+    self.initial_value = initial_value
+    self.initial_mode = initial_mode
+    self.use_default = self.initial_mode == 'constant'
+    self.trainable = trainable
+    self.checkpoint = checkpoint
+
+    self._tables = []
+    self.size_ops = []
+    self.shard_num = len(self.devices)
+
+    assert initial_mode in ["random", "constant"], \
+      "initial mode should be 'constant' or 'random' vs " + initial_mode
+
+    _default_value = _convert_anything_to_init(self.initial_value, self.dim)
+
+    for idx in range(len(self.devices)):
+      with ops.device(self.devices[idx]):
+        _mht = lookup_ops.MutableHashTable(key_dtype=self.key_dtype,
+                                           value_dtype=self.value_dtype,
+                                           default_value=_default_value,
+                                           use_default=self.use_default,
+                                           shared_name="{}-{}of{}".format(
+                                             self.shared_name,
+                                             idx + 1,
+                                             self.shard_num),
+                                           name="{}-{}of{}".format(self.name,
+                                                                   idx + 1,
+                                                                   self.shard_num),
+                                           checkpoint=self.checkpoint)
+
+        self._tables.append(_mht)
+        self.size_ops.append(self._tables[idx].size())
+
+  def upsert(self, keys, values, name=None):
```

Comment:
"Upsert" is not a good choice of wording. Try more idiomatic names such as `InsertOrUpdate`.

Reply:
> "Upsert" is not a good choice of wording. Try more idiomatic names such as `InsertOrUpdate`.
"upsert" is borrowed from DBMS field. You can check here: [https://en.wiktionary.org/wiki/upsert](https://en.wiktionary.org/wiki/upsert)

Reply:
> Thanks for the updates. Several issues that may worth considering:
> 
> 1. These APIs might not worth a separate TF sub-package, i.e. `tf.dynamic_embedding`. How about, e.g., `tf.nn.dynamic_embedding_lookup`?
> 2. Try to override methods in subclass and avoid modifying the base class if not necessary.
> 3. `trainable_wrap` is a resource variable, but `dynamic_embedding.Variable` is not (any specific reason?). I confirmed with the author offline that the `dynamic_embedding.Variable` represents the whole embedding layer, while `TrainableWrap` wraps a single value lookup from embedding. Current naming doesn't reflect this semantics.

1. & 2. --I agree with you, and I will fix the code
3. --As we talked offline, in some application scenario, a `de.Variable` may be looked up by multiple `ids`s , so I believe `TrainableWrap` is needed to maintain the one-to-multi releationship.